pkg/pillar/cmd/volumemgr/blob.go

```diff
@@ -0,0 +1,301 @@
+package volumemgr
+
+import (
+	"io"
+	"os"
+	"time"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	v1types "github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	log "github.com/sirupsen/logrus"
+)
+
+// downloadBlob download a blob from a content tree
+// returns whether or not the BlobStatus has changed
+func downloadBlob(ctx *volumemgrContext, objType string, sv SignatureVerifier, blob *types.BlobStatus) bool {
+
+	changed := false
+	// Make sure we kick the downloader and have a refcount
+	if !blob.HasDownloaderRef {
+		AddOrRefcountDownloaderConfig(ctx, objType, *blob)
+		blob.HasDownloaderRef = true
+		changed = true
+	}
+	// Check if we have a DownloadStatus if not put a DownloadConfig
+	// in place
+	ds := lookupDownloaderStatus(ctx, objType, blob.Sha256)
+
+	// we do not have one, or it is expired or not referenced, then
+	// the one we just created is it; set our State to DOWNLOADING and return
+	if ds == nil || ds.Expired || ds.RefCount == 0 {
+		if ds == nil {
+			log.Infof("downloadStatus not found for blob %s", blob.Sha256)
+		} else if ds.Expired {
+			log.Infof("downloadStatus Expired set for blob %s", blob.Sha256)
+		} else {
+			log.Infof("downloadStatus RefCount=0 for blob %s", blob.Sha256)
+		}
+		blob.State = types.DOWNLOADING
+		return true
+	}
+
+	// we have a valid & non-expired & refcounted DownloadStatus
+	// make sure the blob path matches the target path
+	if ds.Target != "" && blob.Path == "" {
+		blob.Path = ds.Target
+		changed = true
+		log.Infof("From ds set FileLocation to %s for blob %s",
+			ds.Target, blob.Sha256)
+	}
+
+	// make sure the blob state matches the actual state
+	if blob.State != ds.State {
+		blob.State = ds.State
+		changed = true
+	}
+	if ds.Pending() {
+		log.Infof("lookupDownloaderStatus Pending for blob %s", blob.Sha256)
+		return changed
+	}
+	if ds.HasError() {
+		log.Errorf("Received error from downloader for blob %s: %s",
+			blob.Sha256, ds.Error)
+		blob.SetErrorWithSource(ds.Error, types.DownloaderStatus{},
+			ds.ErrorTime)
+		changed = true
+		return changed
+	}
+	if blob.IsErrorSource(types.DownloaderStatus{}) {
+		log.Infof("Clearing downloader error %s", blob.Error)
+		blob.ClearErrorWithSource()
+		changed = true
+	}
+
+	// At this point, we have a valid DownloadStatus, the BlobStatus state and path
+	// are matched up to the DownloadStatus.
+	// Process the actual state to move to the next stage.
+	switch ds.State {
+	case types.INITIAL:
+		// Nothing to do
+	case types.DOWNLOADING:
+		// Nothing to do
+	case types.DOWNLOADED:
+		// signal verifier to start if it hasn't already; add RefCount
+		if verifyBlob(ctx, objType, sv, blob) {
+			changed = true
+		}
+	}
+	if blob.WaitingForCerts {
+		log.Infof("Waiting for certs for blob %s", blob.Sha256)
+		return changed
+	}
+	log.Infof("Waiting for download for blob %s", blob.Sha256)
+	return changed
```

Comment:
If the state became downloaded, this log message doesn't match.

Reply:
Good point. Copy-paste error. Fixing.